### PR TITLE
polygon/sync: Fix `onMilestoneEvent` crash

### PR DESCRIPTION
**Pull request by shohamc1**
Crash was due to a missing early return.

### CODE REVIEW

**Comment by taratorio** [comment]
@shohamc1 please attach more info to the PR - ie the error message in this case. This is good if we have to retrospectively go back via PRs to figure out why some changes were made etc.

File: polygon/sync/sync.go [hunk 1]
```diff
@@ -144,6 +123,36 @@ func (s *Sync) onMilestoneEvent(
...
+			syncLogPrefix("onMilestoneEvent: local chain tip does not match the milestone, unwinding to the previous verified milestone"),
+			"err", err,
+		)
+
+		if err := s.handleMilestoneMismatch(ctx, ccBuilder); err != nil {
+			return err
+		}
+	}
+
+	if err = ccBuilder.Prune(milestone.EndBlock().Uint64()); err != nil {
```
Comment (taratorio) [inline polygon/sync/sync.go:149]: can be simplified to just `return ccBuilder.Prune(milestone.EndBlock().Uint64())`

Comment (taratorio) [reply polygon/sync/sync.go:149]: (ie no need for err check if stmt)

File: polygon/sync/sync.go [hunk 2]
```diff
@@ -144,6 +123,36 @@ func (s *Sync) onMilestoneEvent(
...
+
+	milestoneHeaders := ccBuilder.HeadersInRange(milestone.StartBlock().Uint64(), milestone.Length())
+	err := s.milestoneVerifier(milestone, milestoneHeaders)
+	if err != nil {
+		s.logger.Debug(
+			syncLogPrefix("onMilestoneEvent: local chain tip does not match the milestone, unwinding to the previous verified milestone"),
+			"err", err,
+		)
+
+		if err := s.handleMilestoneMismatch(ctx, ccBuilder); err != nil {
```
Comment (taratorio) [inline polygon/sync/sync.go:144]: I think we may need to return here (even when no err), so can just do `return s.s.handleMilestoneMismatch`

**Review by taratorio — APPROVED** [review]

**Review by antonis19 — APPROVED** [review]

File: polygon/sync/sync.go [hunk 3]
```diff
@@ -144,6 +123,30 @@ func (s *Sync) onMilestoneEvent(
 	return nil
 }
 
+func (s *Sync) onMilestoneEvent(
```
[the remark this answers is not in the document]
Comment (taratorio) [reply polygon/sync/sync.go:126]: I agree. I'm also not a fan of the `onXYZ` naming here. Some suggestions:
- `applyMilestone` - it verifies milestone matches tip & prunes canonical chain builder root if we match the milestone or it unwinds our tip if we don't match the milestone
- `applyNewBlock` - tries to connect blocks, fill gaps, update fork choice if tip changed
- `applyNewBlockHashes` - same as above but using block hashes event

File: polygon/sync/sync.go [hunk 4]
```diff
@@ -132,7 +111,7 @@ func (s *Sync) onMilestoneEvent(
 		return err
 	}
 	if newTip == nil {
-		return errors.New("sync.Sync.onMilestoneEvent: unexpected to have no milestone headers since the last milestone after receiving a new milestone event")
+		return errors.New("sync.Sync.handleMilestoneMismatch: unexpected to have no milestone headers since the last milestone after receiving a new milestone event")
```
[the remark this answers is not in the document]
Comment (taratorio) [reply polygon/sync/sync.go:114]: Agree, does not hurt to log, or wrap the error in a new error with extra info using `fmt.Errorf("%w: id=%d, start=%d, end=%d", errors.New("...existing err..."), milestone.Id, milestone.Start, milestone.End, ....other info....)`. Happy with either if you want to do it in a follow up.

File: polygon/sync/sync.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (taratorio) [reply polygon/sync/sync.go:126]: `applyMilestoneOnTip`, `applyNewBlockOnTip`, `applyNewBlockHashesOnTip` - if we want to be extra verbose (maybe not needed as it is clear? I don't mind either)

[the remark this answers is not in the document]
Comment (taratorio) [reply polygon/sync/sync.go:126]: (feel free to think of better names based on the descriptions above)

**Pull request merged**